src/models/event-timeline-set.ts

```diff
@@ -56,6 +56,40 @@ export interface IRoomTimelineData {
     liveEvent?: boolean;
 }
 
+export interface IAddLiveEventOptions {
+    /** Applies to events in the timeline only. If this is 'replace' then if a
+     * duplicate is encountered, the event passed to this function will replace
+     * the existing event in the timeline. If this is not specified, or is
+     * 'ignore', then the event passed to this function will be ignored
+     * entirely, preserving the existing event in the timeline. Events are
+     * identical based on their event ID <b>only</b>. */
+    duplicateStrategy?: DuplicateStrategy;
+    /** Whether the sync response came from cache */
+    fromCache?: boolean;
+    /** The state events to reconcile metadata from */
+    roomState?: RoomState;
+    /** Whether the timeline was empty before the marker arrived in
+     *  the room. This could be happen in a variety of cases:
+     *  1. From the initial sync
+     *  2. It's the first state we're seeing after joining the room
+     *  3. Or whether it's coming from `syncFromCache` */
+    timelineWasEmpty?: boolean;
+}
```

Comment:
Instead of dealing with unwieldy long function signatures with unlabeled and optional arguments in the usage, I've refactored some of these into option objects now that I am adding yet another `boolean` here (adding `fromInitialState`).

Before:

```js
room.addLiveEvents(mainTimelineEvents, null, fromCache, timelineWasEmpty);
```

After:

```js
room.addLiveEvents(mainTimelineEvents, {
    fromCache,
    fromInitialState: timelineWasEmpty
});
```

Is it acceptable to make the refactor? Do we have any external API consumers that rely on those function signatures not changing?

If necessary, this refactor can be split out to another PR with some effort.

Reply:
> Is it acceptable to make the refactor? Do we have any external API consumers that rely on those function signatures not changing?

Yes we do, this is a library used by more than just Element, so this would be a breaking change so please mark it as such (via the issue label) so the release process knows to bump the version. Or better yet, make it backwards compatible via an overload.

Reply:
> Or better yet, make it backwards compatible via an overload.

Added some overloads 👍 but it would be good to make sure these get a second set of eyes to make sure these look sane 👀

```
eventTimelineset.addLiveEvent
eventTimelineSet.addEventToTimeline
eventTimeline.addEvent
room.addLiveEvent
room.addLiveEvents
```